staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -532,13 +572,35 @@ type WaitFunc func(done <-chan struct{}) <-chan struct{}
 // "uniform pseudo-random", the `fn` might still run one or multiple time,
 // though eventually `WaitFor` will return.
 func WaitFor(wait WaitFunc, fn ConditionFunc, done <-chan struct{}) error {
+	ctx, cancel := contextForChannel(done)
+	defer cancel()
+	return WaitForWithContext(ctx, wait, fn.WithContext())
+}
+
+// WaitForWithContext continually checks 'fn' as driven by 'wait'.
+//
+// WaitForWithContext gets a channel from 'wait()'', and then invokes 'fn'
+// once for every value placed on the channel and once more when the
+// channel is closed. If the channel is closed and 'fn'
+// returns false without error, WaitForWithContext returns ErrWaitTimeout.
+//
+// If 'fn' returns an error the loop ends and that error is returned. If
+// 'fn' returns true the loop ends and nil is returned.
+//
+// context.Canceled will be returned if the ctx.Done() channel is closed
+// without fn ever returning true.
+//
+// When the ctx.Done() channel is closed, because the golang `select` statement is
+// "uniform pseudo-random", the `fn` might still run one or multiple times,
+// though eventually `WaitForWithContext` will return.
+func WaitForWithContext(ctx context.Context, wait WaitFunc, fn ConditionWithContextFunc) error {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 	c := wait(stopCh)
```

Comment:
This call should take a context.

Reply:
I explained here - https://github.com/kubernetes/kubernetes/pull/101668/commits/95f981159a560fb446d05705e0fa451e4d46b437#r626153530

If we want to extend `WaitFunc` to take context, I would recommend a separate PR, thoughts?

Reply:
That link isn't taking me to a comment.  I don't understand why we would stop halfway like this.